src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISExpressDeployer.cs

```diff
@@ -388,6 +398,55 @@ public override void Dispose()
             }
         }
 
+        private class WindowsNativeMethods
+        {
+            [DllImport("user32.dll")]
+            internal static extern IntPtr GetTopWindow(IntPtr hWnd);
+            [DllImport("user32.dll")]
+            internal static extern IntPtr GetWindow(IntPtr hWnd, uint uCmd);
+            [DllImport("user32.dll")]
+            internal static extern uint GetWindowThreadProcessId(IntPtr hwnd, out uint lpdwProcessId);
+            [DllImport("user32.dll")]
+            internal static extern bool PostMessage(HandleRef hWnd, uint Msg, IntPtr wParam, IntPtr lParam);
+        }
+
+        private static void SendStopMessageToProcess(int pid)
+        {
+            for (var ptr = WindowsNativeMethods.GetTopWindow(IntPtr.Zero); ptr != IntPtr.Zero; ptr = WindowsNativeMethods.GetWindow(ptr, 2))
+            {
+                uint num;
+                WindowsNativeMethods.GetWindowThreadProcessId(ptr, out num);
+                if (pid == num)
+                {
+                    var hWnd = new HandleRef(null, ptr);
+                    WindowsNativeMethods.PostMessage(hWnd, 0x12, IntPtr.Zero, IntPtr.Zero);
+                    return;
+                }
+            }
+        }
+
+        private void GracefullyShutdownProcess(Process hostProcess)
+        {
+            if (hostProcess != null && !hostProcess.HasExited)
+            {
+                // Calling hostProcess.StandardInput.WriteLine("q") with StandardInput redirected
+                // for the process does not work when stopping IISExpress
+                // Also, hostProcess.CloseMainWindow() doesn't work either.
+                // Instead we have to send WM_QUIT to the iisexpress process via pInvokes.
+                // See: https://stackoverflow.com/questions/4772092/starting-and-stopping-iis-express-programmatically
+
+                SendStopMessageToProcess(hostProcess.Id);
+                if (!hostProcess.WaitForExit((int)ShutdownTimeSpan.TotalMilliseconds))
+                {
+                    throw new Exception($"iisexpress Process {hostProcess.Id} failed to gracefully shutdown.");
```

Comment:
Don't throw `Exception`

Reply:
I disagree, we really should make the test fail if it didn't gracefully shutdown.

Reply:
Also, check error code as part of graceful shutdown.

Reply:
Don't throw an exception of type `Exception` throw `InvalidOperationException` or something else.